tests/net/adversary.rs

```diff
@@ -7,14 +7,291 @@
+//! ## Adversary model
+//!
+//! The adversary is assumed to have the following capabilities:
+//!
+//! 1. Manipulation of the order in which messages are delivered.
+//! 1. Eavesdropping on any message on the wire, regardless of sender or receiver.
+//! 1. Full control over any node marked as faulty.
+//!
+//! As a consequence, injecting arbitrary messages from faulty nodes into the network is possible,
+//! by sending a message and re-ordering it.
+//!
+//! The following capabilities are explicitly not included:
+//!
+//! 1. Dropping of messages. The networking layer is expected to ensure that no messages
+//!    are lost. A node that drops messages regardless is considered faulty in real-world
+//!    deployments.
+//! 1. Forging message senders. The networking layer is also expected to sign messages and ensure
+//!    that they are not forged.
+//!
+//! ## Handles
+//!
+//! The adversary manipulates the network and nodes exclusively through handles that ensure they do
+//! not violate the constraints defined above. Handles are either mutable or immutable and can, in
+//! some cases be upgraded to actual references, if the underlying node is faulty (see
+//! `NodeHandle::node()` and `NodeHandle::node_mut()`).
 
-use net::{CrankError, NetMessage, VirtualNet};
+use std::cmp;
 
 use hbbft::messaging::{DistAlgorithm, Step};
+use net::{CrankError, NetMessage, Node, VirtualNet};
+
+/// Immutable network handle.
+///
+/// Allows querying public information of the network or getting immutable handles to any node.
+pub struct NetHandle<'a, D: 'a>(&'a VirtualNet<D>)
+where
+    D: DistAlgorithm;
+
+impl<'a, D: 'a> NetHandle<'a, D>
+where
+    D: DistAlgorithm,
+{
+    /// Returns a node handle iterator over all nodes in the network.
+    #[inline]
+    pub fn nodes(&self) -> impl Iterator<Item = NodeHandle<D>> {
+        self.0.nodes().map(NodeHandle::new)
+    }
+
+    /// Returns an iterator over all faulty in the network.
+    ///
+    /// Instead of a handle, returns the node directly, as the adversary gets full access to all
+    /// nodes in the network.
+    #[inline]
+    pub fn faulty_nodes(&self) -> impl Iterator<Item = &Node<D>> {
+        // FIXME: Add an API to handle a step?
+        // Not wrapped in a `NodeHandle`, the adversary gets full access to their own nodes.
+        self.0.faulty_nodes()
+    }
+
+    /// Returns a node handle iterator over all correct nodes in the network.
+    #[inline]
+    pub fn correct_nodes(&self) -> impl Iterator<Item = NodeHandle<D>> {
+        self.0.correct_nodes().map(NodeHandle::new)
+    }
+
+    /// Returns an iterator over all messages in the network.
+    #[inline]
+    pub fn messages(&'a self) -> impl Iterator<Item = &'a NetMessage<D>> {
+        self.0.messages()
+    }
+
+    /// Returns a handle to a specific node handle.
+    #[inline]
+    pub fn get(&self, id: D::NodeId) -> Option<NodeHandle<D>> {
+        self.0.get(id).map(NodeHandle::new)
+    }
+}
+
+/// Queue position.
```

Comment:
Maybe a better doc would be something like: "position in the network's message queue"?